blockchain/blockchain.go

```diff
@@ -2362,57 +2357,42 @@ type BadBlockArgs struct {
 
 // BadBlocks returns a list of the last 'bad blocks' that the client has seen on the network
 func (bc *BlockChain) BadBlocks() ([]BadBlockArgs, error) {
-	blocks := make([]BadBlockArgs, 0, bc.badBlocks.Len())
-	for _, hash := range bc.badBlocks.Keys() {
-		hashKey, ok := hash.(common.CacheKey)
-		if !ok {
-			logger.Error("invalid key type", "expect", "common.CacheKey", "actual", reflect.TypeOf(hash))
-			continue
-		}
-
-		if blk, exist := bc.badBlocks.Peek(hashKey); exist {
-			cacheGetBadBlockHitMeter.Mark(1)
-			block := blk.(*types.Block)
-			blocks = append(blocks, BadBlockArgs{block.Hash(), block})
-		} else {
-			cacheGetBadBlockMissMeter.Mark(1)
-		}
+	blocks, err := bc.db.ReadAllBadBlocks()
+	if err != nil {
+		return nil, err
+	}
+	badblockargs := make([]BadBlockArgs, len(blocks))
+	for _, block := range blocks {
+		hash := block.Hash()
+		badblockargs = append(badblockargs, BadBlockArgs{
+			Hash:  hash,
+			Block: block,
+		})
 	}
-	return blocks, nil
+	return badblockargs, err
 }
 
 // istanbul BFT
 func (bc *BlockChain) HasBadBlock(hash common.Hash) bool {
-	return bc.badBlocks.Contains(hash)
-}
-
-// addBadBlock adds a bad block to the bad-block LRU cache
-func (bc *BlockChain) addBadBlock(block *types.Block) {
-	bc.badBlocks.Add(block.Header().Hash(), block)
+	badBlock := bc.db.ReadBadBlock(hash)
+	if badBlock != nil {
+		return true
+	}
+	return false
 }
 
 // reportBlock logs a bad block error.
 func (bc *BlockChain) reportBlock(block *types.Block, receipts types.Receipts, err error) {
 	badBlockCounter.Inc(1)
-	bc.addBadBlock(block)
+	bc.db.WriteBadBlock(block)
 
 	var receiptString string
 	for i, receipt := range receipts {
 		receiptString += fmt.Sprintf("\t %d: tx: %v status: %v gas: %v contract: %v bloom: %x logs: %v\n",
 			i, receipt.TxHash.Hex(), receipt.Status, receipt.GasUsed, receipt.ContractAddress.Hex(),
 			receipt.Bloom, receipt.Logs)
 	}
-	logger.Error(fmt.Sprintf(`
-########## BAD BLOCK #########
-Chain config: %v
-
-Number: %v
-Hash: 0x%x
-%v
-
-Error: %v
-##############################
-`, bc.chainConfig, block.Number(), block.Hash(), receiptString, err))
+	logger.Error(fmt.Sprintf(`########## BAD BLOCK ######### Chain config: %v Number: %v Hash: 0x%x %v Error: %v`, bc.chainConfig, block.Number(), block.Hash(), receiptString, err))
```

Comment:
Shouldn't this be something like below for better parsing?
```
logger.Error("########## BAD BLOCK #########", "chainConfig", bc.chainConfig, "number", block.Number(), "hash", block.Hash(), "receipt", receiptString, "err", err))
```

Reply:
This should be good though. But `receiptString` is often returned as an empty space. 
So it might seem strange if it's empty. How do you think?

Reply:
I think it's ok because empty strings are distinguishable (i.e., `receipt= err=...`)
My initial thought was this was for the sake of string parsability.

Reply:
@ian0371's suggestion seems better in my view. If the string is empty, the identification of empty is more easy thanks to the header (`receipt` in the suggested log)

Reply:
@ian0371 @hyunsooda Got it. Apply your suggestion.

Reply:
@hqjang-pepper  Actually, I meant logging without `Sprintf()`, but passing variables in the logger's format, i.e., `logger.Error("message", "var1", var1)`.

Reply:
Please check out the suggested log again. The quality of readability seems not improved with formatted logging by `fmt.Sprintf`